### ISSUE TITLE
different metrics for the same destination

### ISSUE THREAD

**Issue by Lotterleben** [Lotterleben/AODVv2-Draft#7]
[Jiazi]

> section 6.4.  Interaction with the Forwarding Plane
> 
>    o  A route has been updated: update the corresponding route in the
> 
> > ```
> >   Routing Information Base.
> > ```
> 
> Do we need to pay attention to the "multi-metrics" issue here? To support
> multi-metrics, do we need special fields in the RIB? If so, we need to call
> it out in the applicability statement.

[Lotte]

> I’d guess it depends where you’re choosing one route over the other? If
> the RIB supports multiple metrics and can sort it out that’s great, if not
> you might want to decide which route to propagate on to the RIB based on
> its metric, or you might want to „translate“ metrics into differently
> prioritized routes in the RIB? (if that makes sense, it should probably be
> mentioned in 4.5.  Local Route Set though)

[Jiazi]
I don't have a solution by hand yet. But as mentioned in my previous
comments, having multiple entries to the same destination can be dangerous.
For example, a packet requires to be routed using metric type A. Router X
has a route supporting metric type A, so it forwards the packet. When the
packet arrives at Router Y .... woops, the route with metric type A is
invalid, but Router Y has a route with metric type B... should it be used?
A safer way is, of course, discarding the packet to avoid loops -- but is
it acceptable?

**Comment by charliep51** [comment]
Hello Lotte,

I agree the packet should be dropped.

Regards,
Charlie P.

**Comment by Lotterleben** [comment]
If @pv0 and @ratliffstan agree I think we can mark this as closed then... :)

**Comment by pv0** [comment]
Agreed (avoiding the whole issue of how you determine which metric type should be used). Perhaps add clarification that by "A route has been updated" and "corresponding route in the RIB" we mean "an AODVv2 route (addr, prefixlen, metrictype) has been updated - and we might have installed a route in the RIB based on that specific AODVv2 route, if so - update it"

**Comment by pv0** [comment]
Quick question - the entry in the RIB wont have MetricType on it will it? The MetricType lives in the AODVv2 LocalRoute. Or is that implementation specific and out of scope? If we assume the RIB is unaware of MetricType, then when a packet arrives, the forwarding process simply looks to see if there's a route, and there is, so it forwards the packet. How would it decide to drop it? The issues Jiazi presents above are valid. Unless we limit to one metric type in use in the network at one time. 

I've updated the text to hopefully be more clear on what exactly is the "route updated" and the "corresponding route".

**Comment by Lotterleben** [comment]
> Quick question - the entry in the RIB wont have MetricType on it will it? The MetricType lives in the AODVv2 LocalRoute. Or is that implementation specific and out of scope? If we assume the RIB is unaware of MetricType, then when a packet arrives, the forwarding process simply looks to see if there's a route, and there is, so it forwards the packet. How would it decide to drop it? The issues Jiazi presents above are valid. Unless we limit to one metric type in use in the network at one time.

Hmmm, true. I suppose it is somewhat implementation specific whether a RIB knows about metric types, but even if it did, data packets that are being forwarded don't know anything about Metric types, so in order to detect the conflicting metrics we'd have to keep track of what was the last hop, and we can't possibly expect the RIB to do that... :/ Not sure what to do about that though.

**Comment by ratliffstan** [comment]
I think the right way to go with this is that the metric evaluation is AODVv2's problem, and is an issue during RREQ/RREP processing (specifically RREP, I guess). Assume the RIB insert doesn't know and doesn't care about metrics. 

Just to make sure I've got this, I think we're talking about the following case: 

Assume that AODVv2 router "A" is rolling along, using metric "<foo>" for all of it's RIB calculations. Due to motion, AODVv2 router "B" becomes a neighbor of "A". Questions to be answered are: 
1. Is it theoretically possible that router "B" is configured to use metric "<bar>" instead of "<foo>"? 
2. If the answer to #1 is "yes", then when "A" receives an RREP from "B", will that RREP contain B's notion of the metric based on "<bar>" instead of "<foo>"? 
3. Does "A" even know that "B" is using a different metric? If so, how would "A" add "B's" metric into a calculation? 
4. In this case, if "A" and "B" are using different metrics, how does "A" determine what is the "best route" (given that "A" gets multiple answers back from the RREQ)?

If router "A" is using metric "<foo>", *and only <foo>*, and router "B" is using metric "<bar>", *and only <bar>*, then you're hosed. You've got no way of calculating an aggregate metric, without making assumptions (like, a default value of the metric). 

It's easier if metrics are used *to calculate a cost*, and the cost is kept around. Then defaults make more sense. But you need to call out how you take the metric, and derive the cost (e.g., you've got to document the equation you use to calculate cost from a set of metrics). 

All of that said - I think we state that we use hop count only for now. Use of multiple metrics is out of scope, and is left for further work.

**Comment by charliep51** [comment]
I reviewed the modifications to the text from the recent pull request and they seem accurate and very well done.

**Comment by pv0** [comment]
Stan, Sorry this is so late. In Step 2/3, the MetricType is included in the message so the router knows the message is about a different metric type that it supports, and ignores the message.
The issue is that if the routers talk about the same metric types and happily learn routes of any number of metric types, but have different rules about what routes go in the RIB, then they aren't using the same rules for forwarding and loops can occur.

With your final sentence, does the "future work" also include how to handle that potential looping issue? The draft does not currently limit to only one metric type in use in the network at one time. I think calculating a cost from the metrics might work, then at least there's one standard way of updating the RIB, hopefully avoiding that looping issue.

**Comment by charliep51** [comment]
Hello folks,

I agree that selecting the correct route for a packet for a destination 
with different metric types is out of scope for AODVv2.  At the end of 
Section 6, we could add the following text:

> Implementations that allow multiple routes to the same destination 
> with different metric types MUST ensure that the packets are forwarded 
> to the destination using the route with the proper metric type.  If 
> packets are forwarded from source to destination using routes at 
> intermediate nodes with mixed metric types, routing loops may result.

Or we could just remain silent about it, but I prefer that it should be 
mentioned.

Methods for route selection include ToS bits or DSCP code points, IPv6 
flow label, etc.  The routing protocol has to find the routes, but after 
the route is discovered, the work of the routing protocol is done.

Regards,
Charlie P.
